src/coprocessor/dag/batch/executors/slow_hash_aggr_executor.rs

```diff
@@ -188,63 +181,108 @@ impl<Src: BatchExecutor> AggregationExecutorImpl<Src> for SlowHashAggregationImp
         entities: &mut Entities<Src>,
         mut input: LazyBatchColumnVec,
     ) -> Result<()> {
-        let rows_len = input.rows_len();
-
-        // 1. Calculate which group each src row belongs to.
-        self.states_offset_each_row.clear();
-
+        let context = &mut entities.context;
         let src_schema = entities.src.schema();
-
-        // TODO: Eliminate these allocations using an allocator.
-        let mut group_by_keys = Vec::with_capacity(rows_len);
-        let mut group_by_keys_offsets = Vec::with_capacity(rows_len);
-        for _ in 0..rows_len {
-            group_by_keys.push(Vec::with_capacity(32));
-            group_by_keys_offsets.push(SmallVec::new());
-        }
-        for group_by_exp in &self.group_by_exps {
-            let group_by_result =
-                group_by_exp.eval(&mut entities.context, rows_len, src_schema, &mut input)?;
-            // Unwrap is fine because we have verified the group by expression before.
-            let group_column = group_by_result.vector_value().unwrap();
-            let field_type = group_by_result.field_type();
-            for row_index in 0..rows_len {
-                group_by_keys_offsets[row_index].push(group_by_keys[row_index].len() as u32);
-                group_column.encode(row_index, field_type, &mut group_by_keys[row_index])?;
-            }
-        }
-        // One extra offset, to be used as the end offset.
-        for row_index in 0..rows_len {
-            group_by_keys_offsets[row_index].push(group_by_keys[row_index].len() as u32);
+        let rows_len = input.rows_len();
+        let group_by_len = self.group_by_exps.len();
+        let aggr_fn_len = entities.each_aggr_fn.len();
+
+        // Decode columns with mutable input first, so subsequent access to input can be immutable
+        // (and the borrow checker will be happy)
+        ensure_columns_decoded(&context.cfg.tz, &self.group_by_exps, src_schema, &mut input)?;
+        ensure_columns_decoded(
+            &context.cfg.tz,
+            &entities.each_aggr_exprs,
+            src_schema,
+            &mut input,
+        )?;
+        assert!(self.group_by_results.is_empty());
```

Comment:
Simply call clear() first, then no need to check any more?

Reply:
I did it on purpose. Then, there couldn;t be any invalid `RpnStackNode` in those `Vec`s. (in case someone accesses elements in `group_by_results` after processing one batch and before the next batch.)